### PR TITLE
Validation doc

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we add the Spine [validation user guide](https://spine.io/docs/guides/validation-user-guide.html).

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@alexander-yevsyukov, @armiol, PTAL.

File: docs/guides/validation-user-guide.md [hunk 1]
```diff
@@ -4,4 +4,437 @@ headline: Validation User Guide
...
+        GoogleId google = 2; 
+        TwitterId twitter = 3; 
+    }
+}
+```
+
+In this case one of the fields `UserIdentity.email`, `UserIdentity.google`,
+and `UserIdentity.twitter` must be set.
+
+`(is_required)` option is not yet supported in Dart.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:80]: I would have something like a yellow exclamation mark (!) sign here. This isn't particularly related to the story flow.

Or we may add a table with the supported platforms per option at the end of the guide.

File: docs/guides/validation-user-guide.md [hunk 2]
```diff
@@ -4,4 +4,437 @@ headline: Validation User Guide
...
+
+Other cases call for conditional required fields. In particular, some fields of a message must be
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:82]: `conditionally`? 

I would rephrase that anyway. `In some other cases, a field may be either required or not, depending on the value of another field. E.g.: ...`

File: docs/guides/validation-user-guide.md [hunk 3]
```diff
@@ -4,4 +4,437 @@ headline: Validation User Guide
...
+In case if a required field is missing, the validation error message will explicitly say so.
+However, if you need a specific error message for this field, you can provide it via
+the `(if_missing)` option:
+
+```proto
+import "spine/options.proto";
+
+message PhoneNumber {
+    string digits = 1 [(required) = true,
+                       (msg_format = "Phone number must contain digits.")];
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:133]: I cannot see any examples of this syntax in our `options.proto`. Is this a valid case?

File: docs/guides/validation-user-guide.md [hunk 4]
```diff
@@ -4,4 +4,437 @@ headline: Validation User Guide
...
+referenced via any other options.
+
+If `(goes)` option is used, the error message can be customized with the `(goes).msg_format`
+parameter. Note that the message should contain two "`%s`" insertion points: first for the name of
+the field declaring the option and second for the name of the field targeted by the option.
+
+#### When `(required)` is implicit
+
+In Spine, by convention, if the first declared field of a [Command](../introduction/naming-conventions.html#command-definitions),
+an [Event](../introduction/naming-conventions.html#event-definitions), or an entity state is
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:147]: "an"? 

Let's rephrase this anyway. `When defining the domain commands, events or entity states, we have found to be convenient that the first field of the respective Message is the identifier. It's just easier to read. Therefore, by convention, Spine treats the first fields of such objects as their IDs: (example).` Then you say — `however if for some reason you have declared the repeated field first, ...`.

Notice, the positive straightforward, then its example, then exceptions and complications.

File: docs/guides/validation-user-guide.md [hunk 5]
```diff
@@ -4,4 +4,437 @@ headline: Validation User Guide
...
+
+```proto
+import "spine/options.proto";
+
+message User {
+    
+    PersonName name = 2 [(validate) = true];
+    // ...
+}
+```
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:199]: We will need more usage examples, for now in Java. We could later have tab splits by platform, like Google does in their docs, — when the time comes.

They are much more illustrative along with the explanation, than the explanation alone.

File: docs/guides/validation-user-guide.md [hunk 6]
```diff
@@ -4,4 +4,437 @@ headline: Validation User Guide
...
+    
+    PersonName name = 2 [(validate) = true,
+                         (if_invalid).msg_format = "User name is invalid."];
+    // ...
+}
+```
+
+### Number bounds
+
+For number fields, Spine defines a few options to limit the range of expected values.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:226]: `numeric fields`?

File: docs/guides/validation-user-guide.md [hunk 7]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
...
 ---
-<p class="coming-soon">Coming soon...</p>
+
+Protobuf is a great tool for structuring domain models. Protobuf definitions help us to standardize
+creating [Value Objects](https://martinfowler.com/bliki/ValueObject.html).
+Spine enhances Protobuf with a validation library.
+
+This guide will walk you though the API of Spine validation library. All of the validation features
+described here are currently supported in the Java environment. Many are supported in Dart as well.
+For more info, see the description of individual constraints.
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:14]: What is supported in JavaScript?

File: docs/guides/validation-user-guide.md [hunk 8]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
 bodyclass: docs
 layout: docs
 ---
-<p class="coming-soon">Coming soon...</p>
+
+Protobuf is a great tool for structuring domain models. Protobuf definitions help us to standardize
+creating [Value Objects](https://martinfowler.com/bliki/ValueObject.html).
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:9]: Not only value objects. We also define entity states. We also define signals that are immutable, but not value objects per se.

Comment (alexander-yevsyukov) [reply docs/guides/validation-user-guide.md:9]: I would start with repeating the fact that in Spine projects data model is defined in Protobuf, and which parts of the model are defined using this language. Then say that we need to make sure that a model is consistent with business rules, etc.

File: docs/guides/validation-user-guide.md [hunk 9]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
...
+convention, Spine treats the first fields of such objects as their IDs:
+
+```proto
+import "spine/options.proto";
+
+// The state of the User Aggregate.
+message User {
+    option (entity).kind = AGGREGATE;
+
+    UserId id = 2;
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:165]: The problem with this sample is that we need more imports, and package for `UserId` type. Otherwise, it's going to be misleading for the reader.

File: docs/guides/validation-user-guide.md [hunk 10]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
...
+referenced via any other options.
+
+If `(goes)` option is used, the error message can be customized with the `(goes).msg_format`
+parameter. Note that the message should contain two "`%s`" insertion points: first for the name of
+the field declaring the option and second for the name of the field targeted by the option.
+
+#### When `(required)` is implicit
+
+When defining the domain [Commands](../introduction/naming-conventions.html#command-definitions),
+[Events](../introduction/naming-conventions.html#event-definitions) or entity states, we have found
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:154]: Don't we need a comma before “or entity states”?

File: docs/guides/validation-user-guide.md [hunk 11]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
...
+field can span between 0 and 59, and the `LocalTime.seconds` field can span between 0.0 and 60.0,
+but can never reach 60. Exclusive boundaries are especially powerful for fractional numbers, since,
+mathematically, there is no hard upper limit which a field value can reach.
+
+Usage of the double dot separator ("`..`") between the bounds is necessary. 
+
+<p class="note">
+In some languages, Protobuf unsigned integers are represented by signed language primitives.
+For example, in Java, a `uint64` is represented with a `long`. If a value of a field in Java will
+overflow into `long` negatives, it will be considered a negative by the validation library. Keep
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:294]: Please capitalise “Validation”, which is the name of the library.

File: docs/guides/validation-user-guide.md [hunk 12]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
...
+a `map` field to only contain non-duplicating elements (values in case of `map`s).
+
+Example:
+
+```proto
+import "spine/options.proto";
+
+// The state of the User Aggregate.
+message User {
+
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:379]: This sample is not a complete definition of the `User` type. So, I think we should have `// ...` in the middle.

File: docs/guides/validation-user-guide.md [hunk 13]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
...
+The definition of `User` itself need not change.
+
+Note that the fields of an external constraint declaration should replicate the fields of the target
+type. In our example, the `Url` type. If the `Url` type had many fields, only those which need any
+validation should be declared. However, note that if the `Url` type declares any validation on its
+own, all of it is discarded and only the "substitute" rules from the `UserPictureConstraint` are
+used.
+
+<p class="warning">
+External constraints are yet supported in Dart.
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:456]: Are NOT yet?

File: docs/guides/validation-user-guide.md [hunk 14]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
 bodyclass: docs
 layout: docs
 ---
-<p class="coming-soon">Coming soon...</p>
+
+Protobuf is a great tool for structuring domain models. Protobuf definitions help us to standardize
+creating [Value Objects](https://martinfowler.com/bliki/ValueObject.html).
+Spine enhances Protobuf with a validation library.
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:10]: I don't think we extend the language or the library. I'd say that the framework provide the Validation library which builds on top of Protobuf options and plugins to Protobuf compiler.

File: docs/guides/validation-user-guide.md [hunk 15]
```diff
@@ -4,4 +4,461 @@ headline: Validation User Guide
...
+
+This guide will walk you though the API of Spine validation library. All of the validation features
+described here are currently supported in the Java environment. Many are supported in Dart as well.
+For more info, see the description of individual constraints. 
+
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:15]: We need a short overview of the goals of the Validation library. Why do people would want to use it?

Then we need to describe how it's done in general. Then go to options.

File: docs/guides/validation-user-guide.md (hunk 7 above)
Comment (dmdashenkov) [reply docs/guides/validation-user-guide.md]: We do not have validation in JavaScript.

**Comment by dmdashenkov** [comment]
@armiol, @alexander-yevsyukov, PTAL again.

File: docs/guides/validation-user-guide.md [hunk 16]
```diff
@@ -151,19 +166,21 @@ the field declaring the option and second for the name of the field targeted by
...
 When defining the domain [Commands](../introduction/naming-conventions.html#command-definitions),
-[Events](../introduction/naming-conventions.html#event-definitions) or entity states, we have found
+[Events](../introduction/naming-conventions.html#event-definitions), or entity states, we have found
 to be convenient that the first field of the respective Message is the identifier. Therefore, by
 convention, Spine treats the first fields of such objects as their IDs:
 
 ```proto
 import "spine/options.proto";
+import "spine/core/user_id.proto";
+import "spine/people/person_name,.proto";
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:176]: Please remove the comma before `.proto`.

File: docs/guides/validation-user-guide.md [hunk 17]
```diff
@@ -197,18 +216,31 @@ In order to enable message field checks, use `(validate)` option:
 
 ```proto
 import "spine/options.proto";
+import "spine/people/person_name,.proto";
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:219]: Please remove the comma before `.proto`.

File: docs/guides/validation-user-guide.md [hunk 18]
```diff
@@ -222,13 +254,14 @@ used:
...
```
Comment (alexander-yevsyukov) [inline docs/guides/validation-user-guide.md:257]: Please remove the comma before `.proto`.

**Comment by dmdashenkov** [comment]
@alexander-yevsyukov, PTAL again.

File: docs/guides/validation-user-guide.md [hunk 19]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
 bodyclass: docs
 layout: docs
 ---
-<p class="coming-soon">Coming soon...</p>
+
+Spine uses Protobuf for defining domain models. It also helps us to standardize creating
+[Value Objects](https://martinfowler.com/bliki/ValueObject.html). Both those usages require
+something more than just a data structure. One of the commodities required for describing domain
+specifics is data validation. Spine builds on top of Protobuf with a Validation library.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:11]: I would rephrase `Spine builds on top of Protobuf with a Validation library.` with more specifics.

```
Spine extensively uses the Validation library when working with any Protobuf messages.
```

File: docs/guides/validation-user-guide.md [hunk 20]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+
+The purpose of the Validation library is:
+ - giving framework users an easy way to maintain simple domain invariants;
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:14]: Let's extract the common parts of the points.

I also think that mentioning Anticorruption layers and domain invariants is a bit too much. Both these statements require the reader to make several leaps from this guide through their experience towards realizing what the author meant in the first place.

File: docs/guides/validation-user-guide.md [hunk 21]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+described here are currently supported in the Java environment. Many are supported in Dart as well.
+For more info, see the description of individual constraints.
+
+## Validation options
+### Required fields
+
+In Protobuf 2 the concept of required fields used to be built into the language. This proved to be
+a [design mistake](https://stackoverflow.com/a/31814967/3183076) and in Protobuf 3 all the fields
+are optional. However, we've revived the concept in Spine Validation library. Unlike the Protobuf 2
+`required` fields, our `(required)` fields do not effect message serialization and thus can be
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:27]: I think `affect`.

File: docs/guides/validation-user-guide.md [hunk 22]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+ignored. Rather than guarding a technical invariant, `(required)` guards a domain model invariant.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:28]: Same as above. There are three unrelated points waiting for the reader to understand themselves:

* Protobuf 2 had this feature, but it was their mistake.
* We do it right. Your serialization/deserialization is safe.
* There are some invariants that need to be guarded. Whether you do that or not — it's your responsibility.

I think you would say this in a much simpler way if you were giving a speech to, let's say, Junior devs.

File: docs/guides/validation-user-guide.md [hunk 23]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+The choice whether or not to use validation lies solely on the developer.
+
+#### How required fields work
+
+Fields in Protobuf may have either a primitive type or a user-defined type. A user-defined type is
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:33]: Previously I commented on the flow "simple definition" -> "simple positive example" -> "more details" -> "another example" -> "negative scenarios and don'ts with more technical stuff".

It is still valid as of now.

File: docs/guides/validation-user-guide.md [hunk 24]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+}
+```  
+
+Here, the field `PhoneNumber.digits` is required. If the API user tries to validate an instance of
+`PhoneNumber` without this field, a `ConstraintViolation` is produced:
+
+```java
+PhoneNumber.newBuilder()
+           .setDigits("")
+           .vBuild(); // ← Throws ValidationException.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:68]: I think we should explicitly tell about `vBuild()` and its difference from `build()`.

We describe the API, and this method is surely its part.

File: docs/guides/validation-user-guide.md [hunk 25]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+    uint32 millimeters = 2 [(max) = { value: "1000" exclusive: true }];
+}
+```
+
+#### Ranges
+
+The `(range)` option is a shortcut for a combination of `(min)` and `(max)`. A range specifies both
+boundaries for a numeric field. `(range)` is a `string` option. The `(range)` notation allow
+declaring inclusive and exclusive boundaries. A round bracket ("`(`" or "`)`") denotes an exclusive
+boundary and a square bracket ("`[`" or "`]`") — an inclusive boundary.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:297]: `tells that the boundary is an inclusive one`.

File: docs/guides/validation-user-guide.md [hunk 26]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+    float seconds 3 [(range) = "[0 .. 60.0)"];
+}
+```
+
+In the example above, the `LocalTime.hours` field can span between 0 and 23, the `LocalTime.minutes`
+field can span between 0 and 59, and the `LocalTime.seconds` field can span between 0.0 and 60.0,
+but can never reach 60. Exclusive boundaries are especially powerful for fractional numbers, since,
+mathematically, there is no hard upper limit which a field value can reach.
+
+Usage of the double dot separator ("`..`") between the bounds is necessary.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:321]: I would rather say "is mandatory" or "is required".

File: docs/guides/validation-user-guide.md [hunk 27]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+    
+    // ...
+
+    spine.time.ZonedDateTime when_placed = 12 [(when).in = PAST];
+    spine.time.ZonedDateTime when_expires = 13 [(when).in = FUTURE];
+}
+```
+
+Note that the value is checked in relation to the current server time. In most cases, this should
+not be an issue. However, be aware that using `FUTURE` in Events and entity states may cause
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:393]: This is an absolutely required thing to tell. However, I don't understand why it's bad for immutable Events. And I would tell more about the entity states, as, again, there are some missing pieces for readers between the statement and the consequences.

File: docs/guides/validation-user-guide.md [hunk 28]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+
+```proto
+import "spine/options.proto";
+import "spine/net/email_address.proto";
+
+// The state of the User Aggregate.
+message User {
+    // ...
+
+    repeated spine.net.EmailAddress recovery_emails = 42 [(distict) = true];
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:414]: `distinct`

File: docs/guides/validation-user-guide.md [hunk 29]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+}
+```
+
+### Non-mutable state fields
+
+For entity states, Spine defines a special validation constraint. It is not typically checked in any
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:420]: This one does not tell whether I as a framework user am able to use `set_once` for my own routines or not.

File: docs/guides/validation-user-guide.md [hunk 30]
```diff
@@ -4,4 +4,497 @@ headline: Validation User Guide
...
+
+// The state of the User Aggregate.
+message User {
+    // ...
+
+    spine.net.Url profile_picture = 42;
+}
+```
+
+How do we add validation to the fields inside `Url` so that only the `User.profile_picture` is
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:462]: Let's also show how `Url` looks. We don't expect readers to take their time and go look into the definition themselves, do we?

It makes even more difference, given you speak of `Url` structure further, implying it's a well-known thing.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: docs/guides/validation-user-guide.md (hunk 29 above)
Comment (armiol) [reply docs/guides/validation-user-guide.md:508]: Can I as a Spine user utilize the `(set_once)` myself? If so, then start with what it is and how to use it. And then describe how Spine uses it internally. This is a piece of advanced information, not essential to gain understanding.

File: docs/guides/validation-user-guide.md [hunk 31]
```diff
@@ -4,4 +4,615 @@ headline: Validation User Guide
...
+In Java, you can validate messages against a `set_once` constraint manually. For that, pass the old
+version and the new version of a message to the `Validate.checkValidChange()` method. For example:
+
+```java
+MyMessage old = getMessage();
+MyMessage changed = doSomeStuff(old);
+Validate.checkValidChange(old, changed);
+```
+
+In Dart, there is no support for this feature.
```
Comment (armiol) [inline docs/guides/validation-user-guide.md:543]: Let's mark this paragraph the same way we did earlier.

File: docs/guides/validation-user-guide.md (hunk 11 above)
Comment (dmdashenkov) [reply docs/guides/validation-user-guide.md:412]: We only capitalize "Validation" when in combination with "library" as in _"Validation library"_. The process "validation" is not capitalized.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

I think two first paragraphs describe the motivation behind creating the library as well as the goals pursued by the user. Is there anything else I should add?
